### PR TITLE
Core/logs: added xp info on login & logout

### LINKED ISSUE #25559: Core/logs: Add xp on login / logout log
**Description:**

Imho it can be interesting to find exploits or simple make stats.

INFO  [entities.player.character] Account: ID (IP: xx.xx.xx.xx) Login Character:[name] GUID Full: xxxx Type: Player Low: xxxx Level: xx (xxxxx/xxxxx XP (xxxx XP left))

**Branch(es):**

both

**TC rev. hash/commit:** 

3947e4cb57d4 N/R

**Operating system:** 

Debian 10 x64 N/R

### PR DESCRIPTION
**Pull request by matanshukry**
**Changes proposed:**

-  Added xp info in logout & login logs

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

Ref #25559 

**Tests performed:**

- [x] It builds
- [x] Tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -2217,6 +2217,16 @@ void Player::RemoveFromGroup(Group* group, ObjectGuid guid, RemoveMethod method
...
+uint32 Player::GetXP() const
+{
+    return m_activePlayerData->XP;
+}
+
+uint32 Player::GetXPForNextLevel() const
+{
+    return m_activePlayerData->NextLevelXP;
+}
+
```
Comment (MaxtorCoder) [inline src/server/game/Entities/Player/Player.cpp:2229]: These getters can be directly in the .h file.

File: src/server/game/Entities/Player/Player.h [hunk 2]
```diff
@@ -1093,6 +1093,8 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         void SetGMVisible(bool on);
         void SetPvPDeath(bool on) { if (on) m_ExtraFlags |= PLAYER_EXTRA_PVP_DEATH; else m_ExtraFlags &= ~PLAYER_EXTRA_PVP_DEATH; }
 
+        uint32 GetXP() const;
+        uint32 GetXPForNextLevel() const;
```
Comment (MaxtorCoder) [inline src/server/game/Entities/Player/Player.h:1097]:
```cpp
uint32 GetXP() const { return m_activePlayerData->XP; }
uint32 GetXPForNextLevel() const { return m_activePlayerData->NextLevelXP; }
```

**Review by MaxtorCoder — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Player/Player.cpp (hunk 1 above)
Comment (matanshukry) [reply src/server/game/Entities/Player/Player.cpp:2229]: I usually only use getters in .h files if the fields are structs/primitives (not pointers).
But sure, I'll do that.

Comment (MaxtorCoder) [reply src/server/game/Entities/Player/Player.cpp:2229]: If you look at `m_activePlayerData` getters (or any simple "UF" getters really) they're all in .h 😄

File: src/server/game/Handlers/CharacterHandler.cpp [hunk 3]
```diff
@@ -1351,8 +1351,9 @@ void WorldSession::HandlePlayerLogin(LoginQueryHolder* holder)
     if (pCurrChar->IsGameMaster())
         SendNotification(LANG_GM_ON);
 
-    TC_LOG_INFO("entities.player.character", "Account: %u (IP: %s) Login Character: [%s] (%s) Level: %d",
-        GetAccountId(), GetRemoteAddress().c_str(), pCurrChar->GetName().c_str(), pCurrChar->GetGUID().ToString().c_str(), pCurrChar->getLevel());
+    TC_LOG_INFO("entities.player.character", "Account: %u (IP: %s) Login Character: [%s] (%s) Level: %d, XP: %s/%s (%s left)",
```
Comment (MaxtorCoder) [inline src/server/game/Handlers/CharacterHandler.cpp:1354]:
```suggestion
    TC_LOG_INFO("entities.player.character", "Account: %u (IP: %s) Login Character: [%s] (%s) Level: %d, XP: %u/%u (%u left)",
```

File: src/server/game/Server/WorldSession.cpp [hunk 4]
```diff
@@ -613,8 +613,9 @@ void WorldSession::LogoutPlayer(bool save)
         // e.g if he got disconnected during a transfer to another map
         // calls to GetMap in this case may cause crashes
         _player->CleanupsBeforeDelete();
-        TC_LOG_INFO("entities.player.character", "Account: %u (IP: %s) Logout Character:[%s] (%s) Level: %d",
-            GetAccountId(), GetRemoteAddress().c_str(), _player->GetName().c_str(), _player->GetGUID().ToString().c_str(), _player->getLevel());
+        TC_LOG_INFO("entities.player.character", "Account: %u (IP: %s) Logout Character:[%s] (%s) Level: %d, XP: %s/%s (%s left)",
```
Comment (MaxtorCoder) [inline src/server/game/Server/WorldSession.cpp:616]:
```suggestion
    TC_LOG_INFO("entities.player.character", "Account: %u (IP: %s) Login Character: [%s] (%s) Level: %d, XP: %u/%u (%u left)",
```

**Review by MaxtorCoder — COMMENTED** [review]
Forgot these 2 lol

**Comment by matanshukry** [comment]
Yeah I've already seen those 2 from the tests. I've actually seen it when I first wrote it, but I noticed the line also uses %s for some numbers so I went with it, and it worked.

**Comment by MaxtorCoder** [comment]
> 
> 
> Yeah I've already seen those 2 from the tests. I've actually seen it when I first wrote it, but I noticed the line also uses %s for some numbers so I went with it, and it worked.

Yeah just letting ya know lol

**Review by ghost — APPROVED** [review]
Looks fairly straightforward & good to me.

**Review by MaxtorCoder — APPROVED** [review]

**Comment by Faq** [comment]
Issue is for 335, but ok, Jack wil cherrypick anyway.

**Comment by MaxtorCoder** [comment]
> Issue is for 335, but ok, Jack wil cherrypick anyway.

Shouldn't matter imo hahah, only thing that has to be changed is the UF getter `GetUInt32Value(ACTIVE_PLAYER_FIELD_XP)` if I recall correctly.

**Comment by illfated** [comment]
> Issue is for 3.3.5, but ok, Jack will cherry-pick anyway.

Very well, in that case, it would be appropriate to change "Closes #25559" to "Ref. #25559" in the main PR comment.

**Comment by matanshukry** [comment]
> > Issue is for 3.3.5, but ok, Jack will cherry-pick anyway.
> 
> Very well, in that case, it would be appropriate to change "Closes #25559" to "Ref. #25559" in the main PR comment.

Will do.

**Comment by jackpoz** [comment]
Thanks for the PR :)

File: src/server/game/Entities/Player/Player.h [hunk 5]
```diff
@@ -1093,6 +1093,8 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         void SetGMVisible(bool on);
         void SetPvPDeath(bool on) { if (on) m_ExtraFlags |= PLAYER_EXTRA_PVP_DEATH; else m_ExtraFlags &= ~PLAYER_EXTRA_PVP_DEATH; }
 
+        uint32 GetXP() const { return m_activePlayerData->XP; }
```
Comment (jackpoz) [inline src/server/game/Entities/Player/Player.h:1096]: btw 335 has `return GetUInt32Value(PLAYER_XP)`

Comment (MaxtorCoder) [reply src/server/game/Entities/Player/Player.h:1096]: Wait, so it is `GetUInt32Value`? Man I never understand legacy UF impl. xD

**Pull request merged**